prov/sm2/src/sm2_rma.c

```diff
@@ -0,0 +1,503 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <sys/uio.h>
+
+#include "ofi_iov.h"
+#include "sm2.h"
+#include "sm2_fifo.h"
+#include "sm2_rma.h"
+
+void sm2_rma_handle_remote_error(struct sm2_ep *ep,
+				 struct sm2_xfer_entry *xfer_entry,
+				 struct sm2_sar_ctx *ctx)
+{
+	int ret = 0;
+	if (ctx->op_flags & FI_DELIVERY_COMPLETE &&
+	    0 == (ctx->status_flags & FI_SM2_SAR_STATUS_COMPLETED)) {
+		ret = sm2_write_err_comp(
+			ep->util_ep.tx_cq, (void *) xfer_entry->hdr.context,
+			xfer_entry->hdr.op_flags, 0, xfer_entry->hdr.cq_data,
+			-FI_EREMOTEIO);
+		if (!ret)
+			ctx->status_flags |= FI_SM2_SAR_STATUS_COMPLETED;
+		else
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Error in sm2_write_err_comp\n");
+	}
+	sm2_freestack_push(ep, xfer_entry);
+	if (ctx->msgs_in_flight == 0)
+		sm2_free_rma_ctx(ctx);
+}
+
+void sm2_rma_handle_local_error(struct sm2_ep *ep,
+				struct sm2_xfer_entry *xfer_entry,
+				struct sm2_sar_ctx *ctx, uint64_t err)
+{
+	int ret = 0;
+	if (err == 0)
+		return;
+	ctx->status_flags |= FI_SM2_SAR_ERROR_FLAG;
+
+	if (0 == (ctx->status_flags & FI_SM2_SAR_STATUS_COMPLETED)) {
+		ret = sm2_write_err_comp(ep->util_ep.tx_cq,
+					 (void *) xfer_entry->hdr.context,
+					 xfer_entry->hdr.op_flags, 0,
+					 xfer_entry->hdr.cq_data, err);
+		if (!ret)
+			ctx->status_flags |= FI_SM2_SAR_STATUS_COMPLETED;
+		else
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Error in sm2_write_err_comp\n");
+	}
+	sm2_freestack_push(ep, xfer_entry);
+	if (ctx->msgs_in_flight == 0)
+		sm2_free_rma_ctx(ctx);
+}
+
+void sm2_fill_rma_ctx(struct sm2_ep *ep, const struct fi_msg_rma *msg,
+		      uint32_t op, uint64_t op_flags, sm2_gid_t peer_gid,
+		      struct sm2_sar_ctx *ctx)
+{
+	ctx->ep = ep;
+	ctx->msg.context = msg->context;
+	ctx->msg.rma_iov_count = msg->rma_iov_count;
+	ctx->msg.iov_count = msg->iov_count;
+	ctx->msg.addr = msg->addr;
+	ctx->msg.data = msg->data;
+
+	assert(msg->iov_count <= SM2_IOV_LIMIT);
+
+	memset(ctx->msg.desc, 0, msg->iov_count * sizeof(*msg->desc));
+	if (msg->desc) {
+		memcpy(ctx->msg.desc, msg->desc,
+		       msg->iov_count * sizeof(*msg->desc));
+	}
+	memcpy(ctx->msg.msg_iov, msg->msg_iov,
+	       msg->iov_count * sizeof(*msg->msg_iov));
+	memcpy(ctx->msg.rma_iov, msg->rma_iov,
```

Comment:
because the structs `fi_msg_rma` and an `sm2_rma_msg` have different levels of pointer indirection particularly for their iov and desc arrays.  the sm2_rma_msg type has the array in the structure, while the fi_msg_rma as the array in application memory.